src/compaction.c

```diff
@@ -499,6 +507,56 @@ int RollingMedReadContext(void *contextPtr, RedisModuleIO *io, __unused int encv
     return TSDB_ERROR;
 }
 
+void *RollingAvgCreateContext(__attribute__((unused)) bool reverse, size_t windowSize) {
+    RollingAvgContext *context = (RollingAvgContext *)malloc(sizeof(RollingMedContext));
+    context->count = 0;
+    context->windowSize = windowSize;
+    context->items_queue = carray_new(double, windowSize);
+    context->sum = 0.0;
+
+    for (int i = 0; i < windowSize; i++){
+        carray_push_back(context->items_queue, 0.0);
+    }
+    return context;
+}
+
+void RollingAvgAddValue(void *contextPtr, double value, __attribute__((unused)) timestamp_t ts) {
+    RollingAvgContext *context = (RollingAvgContext *)contextPtr;
+    double* del_ptr = ((double *)context->items_queue->buf) + context->count;
+
+    // remove the oldest value and add the newest to the sum
+    context->sum += value - *del_ptr;
+    context->count = context->count + 1;
+    if(context->count >= context->windowSize)
+        context->count = 0;
```

Comment:
Why are you zeroing here the count?
I think it should be:
    context->count = min(context->count + 1, context->windowSize);
If we are using window size in the finalize function and not the count, we can actually remove the count field.

Reply:
The count is used as an index to the value that should be removed from the sum which is also where the newest value needs to be stored,  the values are not stored in order from 0 to n but from count to n and 0 to count.
For example for a timeseries with values 1,2,3 with a window of 3, we add 4, 4 is added to the total sum, and replaces 1 which is also removed from the sum.

Reply:
Ok, I understand now, but still I think it will be much more clear to use the api of the carray, instead of using it's internal buf field.

Reply:
carray_pop_front removes and returns the oldest item in the array
carray_push_back inserts new item to the end of the array (queue).

Reply:
ok I agree, will update later today